coverage/ctracer/util.h

```diff
@@ -5,25 +5,19 @@
 #define _COVERAGE_UTIL_H
 
 #include <Python.h>
+#include "pythoncapi_compat.h"    // PyFrame_GetLasti()
 
 /* Compile-time debugging helpers */
 #undef WHAT_LOG         /* Define to log the WHAT params in the trace function. */
 #undef TRACE_LOG        /* Define to log our bookkeeping. */
 #undef COLLECT_STATS    /* Collect counters: stats are printed when tracer is stopped. */
 #undef DO_NOTHING       /* Define this to make the tracer do nothing. */
 
-#if PY_VERSION_HEX >= 0x030B00A0
-// 3.11 moved f_lasti into an internal structure. This is totally the wrong way
-// to make this work, but it's all I've got until https://bugs.python.org/issue40421
-// is resolved.
+#if 0x030B00A1 <= PY_VERSION_HEX && PY_VERSION_HEX < 0x030B00B1
+// Compatibility with Python 3.11.0a1-3.11.0a7.
+// Python 3.11.0b1 adds PyFrame_GetLasti().
 #include <internal/pycore_frame.h>
-#define MyFrame_lasti(f)    ((f)->f_frame->f_lasti * 2)
-#elif PY_VERSION_HEX >= 0x030A00A7
-// The f_lasti field changed meaning in 3.10.0a7. It had been bytes, but
-// now is instructions, so we need to adjust it to use it as a byte index.
-#define MyFrame_lasti(f)    ((f)->f_lasti * 2)
-#else
-#define MyFrame_lasti(f)    ((f)->f_lasti)
+#define PyFrame_GetLasti(f)    ((f)->f_frame->f_lasti * 2)
```

Comment:
The version check is off because since 3.11a1, `f->f_frame` refers to the internal `_PyInterpreterFrame`/`_interpreter_frame` struct, not `PyFrameObject`.  Unfortunately I suspect you need to add the `_PyInterpreterFrame_LASTI` macro and use it like how `GetLastI` uses it for 3.11a1-a7, like what CPython does https://github.com/python/cpython/blob/64a54e511debaac6d3a6e53685824fce435c440c/Objects/frameobject.c#L1162.
`_PyInterpreterFrame_LASTI`:
https://github.com/python/cpython/blob/main/Include/internal/pycore_frame.h#L69